bookkeeper-server/src/main/java/org/apache/bookkeeper/client/UpdateLedgerOp.java

```diff
@@ -81,174 +74,100 @@ public UpdateLedgerOp(final BookKeeper bkc, final BookKeeperAdmin admin) {
-                    LOG.error("Exception while updating ledger metadata", ie);
-                    Thread.currentThread().interrupt();
-                    throw new RuntimeException("Exception while updating ledger", ie.getCause());
-                }
-            }
-        });
-        try {
-            // Wait to finish the issued ledgers.
-            updateBookieCb.get();
-        } catch (ExecutionException ee) {
-            throw new IOException("Exception while updating ledger", ee);
-        } catch (InterruptedException ie) {
-            Thread.currentThread().interrupt();
-            throw new IOException("Exception while updating ledger", ie);
-        } finally {
-            executor.shutdown();
-        }
-    }
-
-    private static final class UpdateLedgerCb implements FutureCallback<Void> {
-        final long ledgerId;
-        final AtomicBoolean stop;
-        final AtomicInteger issuedLedgerCnt;
-        final AtomicInteger updatedLedgerCnt;
-        final Set<Long> outstandings;
-        final CountDownLatch syncObj;
-        final UpdateLedgerNotifier progressable;
-
-        public UpdateLedgerCb(long ledgerId, AtomicBoolean stop, AtomicInteger issuedLedgerCnt,
-                AtomicInteger updatedLedgerCnt, Set<Long> outstandings, CountDownLatch syncObj,
-                UpdateLedgerNotifier progressable) {
-            this.ledgerId = ledgerId;
-            this.stop = stop;
-            this.issuedLedgerCnt = issuedLedgerCnt;
-            this.updatedLedgerCnt = updatedLedgerCnt;
-            this.outstandings = outstandings;
-            this.syncObj = syncObj;
-            this.progressable = progressable;
-        }
-
-        @Override
-        public void onFailure(Throwable th) {
-            LOG.error("Error updating ledger {}", ledgerId, th);
-            stop.set(true);
-            finishUpdateLedger();
+                        outstanding.remove(writePromise);
+                    });
         }
 
-        @Override
-        public void onSuccess(Void obj) {
-            updatedLedgerCnt.incrementAndGet();
-            // may print progress
-            progressable.progress(updatedLedgerCnt.get(), issuedLedgerCnt.get());
-            finishUpdateLedger();
-        }
+        CompletableFuture.allOf(outstanding.stream().toArray(CompletableFuture[]::new))
+            .whenComplete((res, ex) -> {
+                    if (ex != null) {
+                        finalPromise.completeExceptionally(ex);
+                    } else {
+                        finalPromise.complete(null);
+                    }
+                });
 
-        private void finishUpdateLedger() {
-            outstandings.remove(ledgerId);
-            if (outstandings.isEmpty() && stop.get()) {
-                LOG.info("Total number of ledgers issued={} updated={}", issuedLedgerCnt.get(), updatedLedgerCnt.get());
-                syncObj.countDown();
+        try {
+            finalPromise.get();
```

Comment:
use `FutureUtils.result(finalPromise)`, which already wrap the logic for handling ExecutionException